docs: introduce AsyncAPI document (#1868)
asyncapi/website: pages/docs/concepts/asyncapi-document/index.md

Furthermore, the AsyncAPI document acts as a communication contract between `receivers` and `senders` within an event-driven system. It specifies the payload content required when a service produces a message and offers clear guidance to the receiver regarding the message's properties.

```YAML
asyncapi: 3.0.0
info:
  title: Cool Example
  version: 0.1.0
channels:
  userSignedUp:
    address: user/signedup
    messages:
      userSignedUp:
        description: An event describing that a user just signed up.
        payload:
          type: object
          additionalProperties: false
          properties:
            fullName:
              type: string
            email:
              type: string
              format: email
            age:
              type: integer
              minimum: 18
operations: 
  userSignedUp:
    action: send
    channel: 
      $ref: '#/channels/userSignedUp'
```

<Remember>

> You might have additional fields depending on the implemented protocol (i.e., MQTT, AMQP, Kafka, etc.).

Author: ```suggestion
You might have additional fields depending on the implemented protocol (i.e., MQTT, AMQP, Kafka, etc.). For example, your AsyncAPI document could have additional fields for [configuring Kafka bindings](https://github.com/asyncapi/bindings/tree/master/kafka#server-binding-object).
```

---
title: 'Introduction'
weight: 50
---

> The AsyncAPI Specification defines a set of files (AsyncAPI documents) required to describe the API of an application. The AsyncAPI document describes the message-driven API in accordance with the AsyncAPI Specification via JSON objects that conform to the JSON standards. YAML, being a superset of JSON, can also be used. It also offers a standardized approach for documenting and describing asynchronous and event-driven systems, clearly defining various objects such as channels and messages. Additionally, users can leverage both the AsyncAPI document and tools like the AsyncAPI Generator for code and documentation generation, enhancing efficiency and consistency in their development processes.

Reviewer: > The AsyncAPI Specification defines a set of files (AsyncAPI documents) required to describe the API of an application. The AsyncAPI document describes the message-driven API

I'm not a fan of this, but I know this is how we have it in spec document 😆 so first we need to fix it there.

it is not that spec defines a set of files. Spec defines a set of fields that one can use to create one or more asyncapi document that describe an API of the app, and such single AsyncAPI document can reference other files that may contain some fields described in he spec

spec says `The AsyncAPI Specification defines a set of files required to describe the API of an application` - and I think it is a "train of thought" that refers to a fact that one AsyncAPI document can consist of multiple files (because of Reference Object concept, basically refers to [this doc](https://www.asyncapi.com/docs/reference/specification/v3.0.0#file-structure)) and not that you create a set of AsyncAPI documents for an app. I hope you know what I mean, not sure how explain it better.

super hard to explain with words

in example: you have an AsyncAPI document like https://github.com/asyncapi/spec/blob/master/examples/social-media/backend/asyncapi.yaml but it also contains references to files like https://github.com/asyncapi/spec/blob/master/examples/social-media/common/servers.yaml

☝🏼 is what spec explanation means by `The AsyncAPI Specification defines a set of files required to describe the API of an application` because in the end it is not just one AsyncAPI document but also another files with some shared fields

when you write `The AsyncAPI Specification defines a set of files (AsyncAPI documents)`, especially the part `(AsyncAPI documents)` kinda suggests that one app always has set of AsyncAPI documents, but there is always one

please write you know what I mean 🙏🏼 😄 as I have a feeling I got into a loop 😅

Author: Here's what I understand... lemme know if you agree @derberg 

The original spec text seems to inaccurately suggest that multiple AsyncAPI documents are required for each application, whereas you're saying that a single AsyncAPI document, potentially referencing other files, is typically used to describe an application's API. 

**Perhaps it is more accurate to say that the AsyncAPI Specification defines a set of fields that can be used in one or more AsyncAPI documents to describe an application's API. A single AsyncAPI document can reference other files, but it doesn't mean that multiple AsyncAPI documents are required for a single application.**

While the AsyncAPI document is the primary file describing the API, it can include references to other files that contain additional details or shared fields. This doesn't equate to requiring multiple AsyncAPI documents for a single application.

## 📣 How about the following instead?

Proposing the following update to the Spec and this document:
> The AsyncAPI Specification defines a set of fields that can be used in an AsyncAPI document to describe an application's API. The document may reference other files for additional details or shared fields, but it is typically a single, primary document that encapsulates the API description.

Author: Spec PR @derberg: https://github.com/asyncapi/spec/pull/1003

Reviewer: perfect, thanks a 💯 you can update this PR to and we can merge it

Author: done!

> The AsyncAPI Specification defines a set of files (AsyncAPI documents) required to describe the API of an application. The AsyncAPI document describes the message-driven API in accordance with the AsyncAPI Specification via JSON objects that conform to the JSON standards. YAML, being a superset of JSON, can also be used. It also offers a standardized approach for documenting and describing asynchronous and event-driven systems, clearly defining various objects such as channels and messages. Additionally, users can leverage both the AsyncAPI document and tools like the AsyncAPI Generator for code and documentation generation, enhancing efficiency and consistency in their development processes.

Author: ```suggestion
The AsyncAPI Specification defines a set of fields that can be used in an AsyncAPI document to describe an application's API. The document may reference other files for additional details or shared fields, but it is typically a single, primary document that encapsulates the API description.
```

---
title: 'Introduction'
weight: 50
---

> 

Author: How about now @smoya? 😸

> The AsyncAPI Specification defines a set of files (AsyncAPI documents) required to describe the API of an application. The AsyncAPI document describes the message-driven API in accordance with the AsyncAPI Specification via JSON objects that conform to the JSON standards. (YAML, being a superset of JSON, can also be used.) It also offers a standardized approach for documenting and describing asynchronous and event-driven systems, clearly defining various components like channels and messages. Additionally, users can leverage both the AsyncAPI document and tools like the AsyncAPI Generator for code and documentation generation, enhancing efficiency and consistency in their development processes.

Reviewer 2: Nitpick: 

```suggestion
The AsyncAPI Specification defines a set of files (AsyncAPI documents) required to describe the API of an application. The AsyncAPI document describes the message-driven API in accordance with the AsyncAPI Specification via JSON objects that conform to the JSON standards. YAML, being a superset of JSON, can also be used. It also offers a standardized approach for documenting and describing asynchronous and event-driven systems, clearly defining various objects such as channels and messages. Additionally, users can leverage both the AsyncAPI document and tools like the AsyncAPI Generator for code and documentation generation, enhancing efficiency and consistency in their development processes.
```

Or removing the brackets maybe? You are the expert so feel free to skip my suggestion.

Author: Thank you, and yes... I will remove the brackets too!

> The AsyncAPI Specification defines a set of files (AsyncAPI documents) required to describe the API of an application. The AsyncAPI document describes the message-driven API in accordance with the AsyncAPI Specification via JSON objects that conform to the JSON standards. (YAML, being a superset of JSON, can also be used.) It also offers a standardized approach for documenting and describing asynchronous and event-driven systems, clearly defining various components like channels and messages. Additionally, users can leverage both the AsyncAPI document and tools like the AsyncAPI Generator for code and documentation generation, enhancing efficiency and consistency in their development processes.

Reviewer 2: Additionally, we use the word `components` here, but we technically call those `objects`: 
![Google Chrome_17R585e6@2x](https://github.com/asyncapi/website/assets/1083296/7378824e-21b5-4c9e-9b00-a2dd5decd5ea)
See https://www.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.17#schema

However, depending on the context we are, we might want to call them components (maybe the user is the first time it reads this?). I'm not sure, being completely honest 🤷

Author: I see, thank you for this detail... I think it makes sense to call them objects, so I have updated that line accordingly. 😄✌🏽

> The AsyncAPI Specification defines a set of files required to describe the API of an application. One of those files is the AsyncAPI document; it describes the message-driven API in accordance with the AsyncAPI Specification via JSON objects that conform to the JSON standards. (YAML, being a superset of JSON, can also be used.) The AsyncAPI document offers a standardized approach for documenting and describing asynchronous and event-driven systems, clearly defining various components like channels and messages. Additionally, users can leverage both the AsyncAPI document and tools like the AsyncAPI Generator for code and documentation generation, enhancing efficiency and consistency in their development processes.

Reviewer 2: >  One of those files is the AsyncAPI document

I don't think we have other kind of files rather than the AsyncAPI document.

Author: ah, noted! fixing

> The AsyncAPI Specification defines a set of files required to describe the API of an application. One of those files is the AsyncAPI document; it describes the message-driven API in accordance with the AsyncAPI Specification via JSON objects that conform to the JSON standards. (YAML, being a superset of JSON, can also be used.) The AsyncAPI document offers a standardized approach for documenting and describing asynchronous and event-driven systems, clearly defining various components like channels and messages. Additionally, users can leverage both the AsyncAPI document and tools like the AsyncAPI Generator for code and documentation generation, enhancing efficiency and consistency in their development processes.

Author: ```suggestion
The AsyncAPI Specification defines a set of files (AsyncAPI documents) required to describe the API of an application. The AsyncAPI document describes the message-driven API in accordance with the AsyncAPI Specification via JSON objects that conform to the JSON standards. (YAML, being a superset of JSON, can also be used.) It also offers a standardized approach for documenting and describing asynchronous and event-driven systems, clearly defining various components like channels and messages. Additionally, users can leverage both the AsyncAPI document and tools like the AsyncAPI Generator for code and documentation generation, enhancing efficiency and consistency in their development processes.
```

> Furthermore, the AsyncAPI document acts as a communication contract between `receivers` and `senders` within an event-driven system. It specifies the payload content required when a service produces a message and offers clear guidance to the receiver regarding the message's properties.

Reviewer 2: Nitpick: I think, the following littel change makes sense for consistency's sake.

```suggestion
Furthermore, the AsyncAPI document acts as a communication contract between `receivers` and `senders` within an event-driven system. It specifies the payload content required when a service sends a message and offers clear guidance to the receiver regarding the message's properties.
```

```YAML
asyncapi: 3.0.0
info:
  title: Cool Example
  version: 0.1.0
channels:
  userSignedUp:
    address: user/signedup
    messages:
      userSignedUp:
        description: An event describing that a user just signed up.
        payload:
          type: object

>           additionalProperties: false

Reviewer 2: Nitpick: I would remove this line unless it is completely needed for your example. It doesn't add value IMHO and I believe can add too much noise to the final user. The example is so simple that this "overcomplicates" it a bit.

```suggestion
```

Author: sure! sounds good to me

          properties:
            fullName:
              type: string
            email:
              type: string
              format: email
            age:
              type: integer
              minimum: 18
operations: 
  userSignedUp:
    action: send
    channel: 

>       $ref: '#/channels/userSignedUp'

Reviewer 2: The example is a very simple example and I like it. 👍 
In fact, it is pretty similar to our "official" simple example: https://github.com/asyncapi/spec/blob/next-major-spec/examples/simple-asyncapi.yml

Author: Well, you wrote it, so I hope you like it! 😅😂😂

---
title: 'Introduction'

> weight: 1

Reviewer: with subsections it works this way that the "starting weight" is the weight of the `_section.md` so change it to 50 or more and all the next documents should follow. This way you fix preview and make sure it shows in the UI

> weight: 1

Author: ```suggestion
weight: 50
```

---

> The AsyncAPI document is a YAML or JSON file that describes your implementation of the AsyncAPI specification. Providing a standardized way to document and describe asynchronous and event-driven systems, the AsyncAPI document generates documentation/code and defines the different components of your event-driven application, such as the events, channels, and message formats.

Reviewer 2: > The AsyncAPI document is a YAML or JSON file that describes your implementation of the AsyncAPI specification. 

I would like to recommend, if possible, sticking with a unique description of what an AsyncAPI document is, at least for the first sentence. Then we can expand further and freely here in this doc. I mean to be aligned with what the spec says about in https://www.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.12#introduction. 

Also related to this, I think is of your interest the issue I recently created regarding [consistency in wording between the terms "Application", "API" and derivatives.](https://github.com/asyncapi/spec/issues/949)

Author: 1. Ah yes, you’re absolutely right about the first sentence being consistent in defining the AsyncAPI Document. Thank you for calling that out, I personally think that I should go more in line with: https://www.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.16#format. To me, this section feels like a clearer intro to the topic... wondering what you think about the new draft approach. Anyway, let me know what you think about my next commit 😅 

```
The files describing the message-driven API in accordance with the AsyncAPI Specification are represented as JSON objects and conform to the JSON standards. YAML, being a superset of JSON, can be used as well to represent an AsyncAPI Specification file.
```

2. Noted, sounds like we should be going for **API** instead of application, per the discussion you linked.

> The AsyncAPI document is a YAML or JSON file that describes your implementation of the AsyncAPI specification. Providing a standardized way to document and describe asynchronous and event-driven systems, the AsyncAPI document generates documentation/code and defines the different components of your event-driven application, such as the events, channels, and message formats.

Reviewer 2: > the AsyncAPI document generates documentation/code {...}

The document doesn't generate documentation or code per see, but through generators and other tooling. I would reword that to clarify it.

>  {...} and defines the different components of your event-driven application, such as the events, channels, and message formats.

I would replace the word `events` by `messages` because events is only one type of messages. For example, can be a request, a command, a query, etc.

Author: Oh got it, right! Fixed both.

> The AsyncAPI document is a YAML or JSON file that describes your implementation of the AsyncAPI specification. Providing a standardized way to document and describe asynchronous and event-driven systems, the AsyncAPI document generates documentation/code and defines the different components of your event-driven application, such as the events, channels, and message formats.

Author: ```suggestion
The AsyncAPI document is a file describing the message-driven API in accordance with the AsyncAPI Specification via JSON objects that conform to the JSON standards. (YAML, being a superset of JSON, can also be used.)
The AsyncAPI document offers a standardized approach for documenting and describing asynchronous and event-driven systems, clearly defining various components like channels and messages. Additionally, users can leverage both the AsyncAPI document and tools like the AsyncAPI Generator for code and documentation generation, enhancing efficiency and consistency in their development processes.
```

> Primarily, it provides a method to produce comprehensive, consistent, and standardized AsyncAPI documentation. Additionally, it serves as a contract for communication between consumers and producers in an event-driven system. It specifies what should be included in the payload when a service produces a message and provides information to the consumer about the properties in the message. 

Reviewer 2: > Primarily, it provides a method to produce comprehensive, consistent, and standardized AsyncAPI documentation. 

I think this is already stated in the previous sentence:

> Providing a standardized way to document and describe asynchronous and event-driven systems

Author: agreed, thanks for calling that out

> Primarily, it provides a method to produce comprehensive, consistent, and standardized AsyncAPI documentation. Additionally, it serves as a contract for communication between consumers and producers in an event-driven system. It specifies what should be included in the payload when a service produces a message and provides information to the consumer about the properties in the message. 

Reviewer 2: Regarding the words `Producer/Subscriber`, we recently changed it to become `Sender/Receiver`. See https://github.com/asyncapi/spec/pull/934/files#diff-34bca0c91901dffce093114c7d4cac16477410fbca7cabf3a1b888d7dfc3eb4dR41

Author: noted, thank you 🙏

> Primarily, it provides a method to produce comprehensive, consistent, and standardized AsyncAPI documentation. Additionally, it serves as a contract for communication between consumers and producers in an event-driven system. It specifies what should be included in the payload when a service produces a message and provides information to the consumer about the properties in the message. 

Author: ```suggestion
Furthermore, the AsyncAPI document acts as a communication contract between `receivers` and `senders` within an event-driven system. It specifies the payload content required when a service produces a message and offers clear guidance to the consumer regarding the message's properties.
```

```YAML
asyncapi: 3.0.0
info:
  title: Cool Example
  version: 0.1.0
channels:
  user/signedup:

>     subscribe:

Reviewer 2: This is actually an invalid document. Version 3 changed how channels, operations, and messages are described and linked. 
It also changed the terms `Publish/Subscribe` to `Receive/Send` **in that order**, since documents changed from being Client-Centric to Application-Centric (meaning they describe the applications API and not the client's interaction within those APIs).

Ok, probably too much info there to digest 😓 .

Let's start with some of the changes in v3 that affect this example:

- `Publish/Subscribe` thing. Here is a simple cheat sheet I made for [the talk I presented at the EDA Summit 23](https://edasummit.com/videos/asyncapi-v3-unleashing-the-power-of-asynchronous-apis-design/) that helps to convert from v2 to v3.
  - ![](https://github.com/asyncapi/website/assets/1083296/6a14c921-6d90-4634-94df-85b4b04c0171)
- Channels now have a new field called `address` that holds the address of the channel. Now the name of the channel can be whatever name. I recommend `userSignedUp` here.
- Channels don't include the Operation anymore. Instead, Operations are defined in the root document and are "linked" to the channel. 

Applying all the previous, your document will look like this:

```yaml
asyncapi: 3.0.0
info:
  title: Cool Example
  version: 0.1.0
channels:
  userSignedUp:
    address: user/signedup
    messages:
      userSignedUp:
        description: An event describing that a user just signed up.
        payload:
          type: object
          additionalProperties: false
          properties:
            fullName:
              type: string
            email:
              type: string
              format: email
            age:
              type: integer
              minimum: 18
operations: 
  userSignedUp:
    action: send
    channel: 
      $ref: '#/channels/userSignedUp'
```

Author: Whoa, this was a lot of info but I really appreciate how fully you described everything. Thank you 🙏 !!!

 One thing slightly confuses me... When you wrote `changed the terms Publish/Subscribe to Receive/Send in that order`, that made complete sense to me. With you so far. But then when I looked at your cheatsheet, my mind feels like it's trying to do gymnastics.  Are you saying that in v3: 
 - applications no longer subscribe, they `send` 
 - applications no longer publish, they `receive`
? 
I have copied your updated code snippet and will integrate in my next commit. ✌🏽

Reviewer 2: The reality is that the approach or point of view of the documented API changes. In v3, we now clearly state that:

> The AsyncAPI Specification defines a set of files required to describe the API of an [application](https://www.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.16#definitionsApplication).

Meaning the document describes the API of the application that offers the AsyncAPI document. 
That means operations with the action `send`, means that the application will perform that operation by sending the messages. On the contrary, `receive` means the application expects to receive those messages.

      message:
        description: An event describing that a user just signed up.
        payload:
          type: object
          additionalProperties: false
          properties:
            fullName:
              type: string
            email:
              type: string
              format: email
            age:
              type: integer
              minimum: 18
```

> ## AsyncAPI document components

Reviewer 2: We refer to these as `Resources` or `Objects` as far as I can see in the spec:

> AsyncAPI Object
This is the root document object for the API specification. It combines resource listing and API declaration together into one document. 
>
> Source: https://www.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.12#schema

I say that because `components` is also a thing in the AsyncAPI document: https://www.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.12#componentsObject and can sound confusing.

Author: Ah, thank you for calling that out!

> The AsyncAPI document follows a specific structure with certain headers or fields:

Reviewer 2: ```suggestion
The AsyncAPI document follows a specific structure with certain fields:
```

- **`info`**: This required field holds metadata about the API. The `info` section must include the following sub-fields:
   
   - `title`: A descriptive title of the API.
   
   - `version`: An API version identifier.
   
   Optional fields within the `info` section could include `description`, `termsOfService`, `contact` (with subfields for `name`, `url`, `email`), and `license` (with `name` and `url` subfields).

- **`servers`**: An optional field specifying the server(s) where the API is available. Each server must have a `url` and may have `protocol`, `protocolVersion`, `description`, `variables`, `security`, and `bindings`.

- **`channels`**: This required field describes the paths available to the API and must include at least one channel. Each channel could include fields such as `description`, `subscribe`, `publish`, `parameters`, and `bindings`.

- **`components`**: An optional section for defining reusable components like `messages`, `schemas`, `securitySchemes`, and others.

- **`tags`**: An optional array of tags used by the specification with additional metadata. Each tag object can have `name` and `description`.

> - **`externalDocs`**: An optional section to provide more information, such as a link to the API’s external documentation with a `description` and a `url`.

Reviewer 2: I'm concerned about the maintainability of this. I agree this won't change a lot in minor versions but can potentially do for major ones. Could we maybe just link to each section of the Spec for each of these?

In fact, this list is from v2. v3 includes more and do some changes in others.

Author: I agree on maintainability, thank you for calling that out. 

As of today, this info is already covered via https://v3.asyncapi.com/docs/concepts/asyncapi-document/structure. Thus, I will remove this section from this document because it would be duplicate content.

<Remember>

> You might have additional fields depending on the implemented protocol (i.e., MQTT, AMQP, Kafka, etc.).

Reviewer 2: I would clarify this with an example

Author: Thank you, noted. Added an example, you let me know if you like it.
